### PR TITLE
Mul 2PC

### PR DESCRIPTION
**Pull request by th4s**
This PR implements the Gilboa OT protocol for getting additive shares of a product.

### CODE REVIEW

File: m2a-2pc/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use rand::{Rng, SeedableRng};
+    use rand_chacha::ChaCha12Rng;
+
+    fn ot_mock(envelopes: ([u128; 128], [u128; 128]), choices: u128) -> [u128; 128] {
+        let mut out = [0_u128; 128];
+        for (k, digit) in out.iter_mut().enumerate() {
+            let mask = (choices >> k) & 1;
```
Comment (themighty1) [inline m2a-2pc/src/lib.rs:49]: shouldnt this be called `bit` instead of `mask`?

File: m2a-2pc/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use rand::{Rng, SeedableRng};
+    use rand_chacha::ChaCha12Rng;
+
+    fn ot_mock(envelopes: ([u128; 128], [u128; 128]), choices: u128) -> [u128; 128] {
+        let mut out = [0_u128; 128];
+        for (k, digit) in out.iter_mut().enumerate() {
```
Comment (themighty1) [inline m2a-2pc/src/lib.rs:48]: i think that `digit` is sth like 1,2,3,4, `number` would be a better name

**Review by themighty1 — APPROVED** [review]
lgtm

**Comment by th4s** [comment]
@sinui0 yep full agree.

File: m2a-2pc/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+        Self(share)
+    }
+
+    /// Return inner share
+    pub fn inner(&self) -> u128 {
+        self.0
+    }
+}
+
+impl From<[u128; 128]> for AddShare {
```
[the remark this answers is not in the document]
Comment (th4s) [reply m2a-2pc/src/lib.rs:63]: alright!

File: m2a-2pc/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+    pub fn inner(&self) -> u128 {
+        self.0
+    }
+
+    /// Turn into an additive share and masked encodings
+    ///
+    /// This function returns
+    ///   * `AddShare` - The sender's additive share; this is `y` in the paper
+    ///   * `MaskedEncoding` - Used for oblivious transfer; t0 and t1 in the paper
+    pub fn encode(&self) -> (AddShare, MaskedEncoding) {
```
[the remark this answers is not in the document]
Comment (th4s) [reply m2a-2pc/src/lib.rs:37]: yepp

**Review by themighty1 — APPROVED** [review]
ack

**Pull request merged**